### DISCUSSION TITLE
Ouath2 flow that depends on another auth flow

### DISCUSSION THREAD

**Discussion by loay106** [fastapi/fastapi#9321] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi.security import OAuth2AuthorizationCodeBearer
    
    oauth2_store = OAuth2AuthorizationCodeBearer(
        authorizationUrl="users/token",
        tokenUrl="users/token"
    )
    
    @app.post("/stores/auth/{store_id}")
    def create_store_token(store_id: str, user: Annotated[User, Depends(get_current_user)]):
        store_token = create_store_token(store_id, user)
        return store_token

    @app.get("/stores/items/{store_id}")
    def get_store_items(store: Annotated[Store, Security(get_current_store, scopes=["store:items"])]):
        return store.items
```

### Description

Hi,

First of all I want to thank you for the great framework you're building, it is intuitive, easy to use and regularly updated.

Issue:

I'm trying to implement a FastAPI application with two authentication/authorization flows:

1. `Users API`: users can create an account and then authenticate (login) with their credentials to receive a JWT `user token`. Users can then send the `user token` to the application and interact with the `Users API`.
2. `Stores API`: users can send the `user token` from above and request a JWT `store token`, which includes the scopes/permissions allowed for them for that specific store. Users can then send the `store token` and interact with `Stores API`.

Following the Advanced security tutorial I was able to implement the auth flow for the `Users API`.

How can I implement the auth flow for the `Stores API`? More specifically, how can I tell FastAPI that to request a `store token`, users need to be authenticated in the `Users API` and the token they receive should be used to get a new access token, for the `Stores API`.

Any help is appreciated! Thank you.

### Operating System

Linux

### FastAPI Version

0.95.0

### Python Version

3.10

**Comment by ebottos94** [comment]
Hi @loay106, I would proceed in this way :

First : Implement your first authorization as docs suggest. 
Second: create another secret key for your new jwt authentication, create an endpoint that check your first auth with ```user token``` and return another token based on your second secret key, this will be your ```store token```. Next implement a function that receives the new token and validate this, all your endpoints in ```Stores API``` will depend by this function. 

I implemented an example, starting from the [example in the documentation](https://fastapi.tiangolo.com/tutorial/security/oauth2-jwt/). The three new functions are : ```create_stores_token```, ```get_stores_token``` and ```get_current_store_user```. The two secret keys are : ```SECRET_KEY_USERS``` and ```SECRET_KEY_STORES```

N.B. to test /user/store endpoint you can't use docs swagger UI, as this is based on your first authorization, so to test this endpoint you need to send a request to this endpoint (for example with postman). 


The code : 

```python
from typing import Annotated
from datetime import datetime, timedelta
from jose import JWTError, jwt
from passlib.context import CryptContext
from fastapi import FastAPI, Depends, status, HTTPException
from fastapi.security import OAuth2PasswordBearer, OAuth2PasswordRequestForm
from pydantic import BaseModel

SECRET_KEY_USERS = "09d25e094faa6ca2556c818166b7a9563b93f7099f6f0f4caa6cf63b88e8d3e7"
SECRET_KEY_STORES = "da355544922131df45a7fdd181df4285799e51e6d50ea18c137cee7b5cc64d58"
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 30

fake_users_db = {
    "johndoe": {
        "username": "johndoe",
        "full_name": "John Doe",
        "email": "johndoe@example.com",
        "hashed_password": "$2b$12$IiIG6dQ94rTIbK7re38a..TUzgBG0ihfm2XCXhY8bgRP0d8h/nZIK",#pwd=test
        "disabled": False,
    }
}


class Token(BaseModel):
    access_token: str
    token_type: str


class TokenData(BaseModel):
    username: str | None = None


class User(BaseModel):
    username: str
    email: str | None = None
    full_name: str | None = None
    disabled: bool | None = None


class UserInDB(User):
    hashed_password: str

pwd_context = CryptContext(schemes=["bcrypt"], deprecated="auto")

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")


app = FastAPI()

def verify_password(plain_password, hashed_password):
    return pwd_context.verify(plain_password, hashed_password)


def get_password_hash(password):
    return pwd_context.hash(password)

def get_user(db, username: str):
    if username in db:
        user_dict = db[username]
        return UserInDB(**user_dict)
    
def authenticate_user(fake_db, username: str, password: str):
    user = get_user(fake_db, username)
    if not user:
        return False
    if not verify_password(password, user.hashed_password):
        return False
    return user

def create_access_token(data: dict, expires_delta: timedelta | None = None):
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=15)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY_USERS, algorithm=ALGORITHM)
    return encoded_jwt


    

async def get_current_user(token: Annotated[str, Depends(oauth2_scheme)]):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY_USERS, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(fake_users_db, username=token_data.username)
    if user is None:
        raise credentials_exception
    return user

@app.post("/token", response_model=Token)
async def login_for_access_token(
    form_data: Annotated[OAuth2PasswordRequestForm, Depends()]
):
    user = authenticate_user(fake_users_db, form_data.username, form_data.password)
    if not user:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Incorrect username or password",
            headers={"WWW-Authenticate": "Bearer"},
        )
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_access_token(
        data={"sub": user.username}, expires_delta=access_token_expires
    )
    return {"access_token": access_token, "token_type": "bearer"}

@app.get("/users/me/", response_model=User)
async def read_users_me(
    current_user: Annotated[User, Depends(get_current_user)]
):
    return current_user




#STORE TOKEN IMPLEMENTATION

def create_stores_token(data: dict, expires_delta: timedelta | None = None):
    to_encode = data.copy()
    if expires_delta:
        expire = datetime.utcnow() + expires_delta
    else:
        expire = datetime.utcnow() + timedelta(minutes=15)
    to_encode.update({"exp": expire})
    encoded_jwt = jwt.encode(to_encode, SECRET_KEY_STORES, algorithm=ALGORITHM)
    return encoded_jwt

async def get_stores_token(token: Annotated[str, Depends(oauth2_scheme)]):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY_USERS, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(fake_users_db, username=token_data.username)
    if user is None:
        raise credentials_exception
    access_token_expires = timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    access_token = create_stores_token(
        data={"sub": user.username}, expires_delta=access_token_expires
    )
    return {"access_token": access_token, "token_type": "bearer"}

async def get_current_store_user(token: Annotated[str, Depends(oauth2_scheme)]):
    credentials_exception = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Could not validate credentials",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY_STORES, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exception
        token_data = TokenData(username=username)
    except JWTError:
        raise credentials_exception
    user = get_user(fake_users_db, username=token_data.username)
    if user is None:
        raise credentials_exception
    return user

@app.get("/token_store", response_model=Token)
async def stores_token_api(
    token_store: Annotated[Token, Depends(get_stores_token)]
):
    return token_store

@app.get("/user/store", response_model=User)
async def read_store_user(
    user: Annotated[User, Depends(get_current_store_user)]
):
    return user
```

**Reply by loay106** [reply]
Hi @ebottos94, this is exactly what I have right now but I was wondering if there's a way to achieve the same result and still have it work with OpenAPI.

I was thinking of creating two FastAPI applications. The first would implement the Users API and the second would implement the Store API.
That way, the users application can be an authentication provider for the store application and it should be possible to implement the auth flow in the store application using the code flow auth as mentioned [here](https://fastapi.tiangolo.com/advanced/security/oauth2-scopes/#about-third-party-integrations)

**Reply by ebottos94** [reply]
For 'the same result' you mean that you want another endpoint with auth form (as /token in the example)? And if yes, why? There are two different user databases ?

**Reply by loay106** [reply]
No, I mean that I want it to work with the Swagger UI as well. I have one users database and one users API.

With the auth flow above once the user authenticates in the Swagger UI the same user token is passed as a store token in the store API, which is not the intended result.

**Reply by ebottos94** [reply]
Yes because in Swagger UI with auth setted on user authentication save relative token in current session and pass this token via header to endpoint that need authentication. But in real applications you will pass the other token via header and so it works!

**Reply by loay106** [reply]
Is this also applied in the generated [client](https://fastapi.tiangolo.com/advanced/generate-clients/) as well?

I know that in my requests to the users or stores API I can specify the parameters that I need to send, however, if I'm planning to add more endpoints to the API's  it will add an overhead of having to re-configure the authentication flow in them every time a change is made and a new client is generated.

Since the clients are generated from the OpenAPI specs, I would like to have multiple oauth2 schemes. However, I'm aware that OpenAPI is limited to one auth scheme.

If I would still like to have two oauth2 schemes in my application, the generated clients and the OpenAPI, what are my options?

One option I can think of is having two FastAPI applications:
The first application implements the users API and the auth flow in the tutorial should be enough to be implement the auth flow. 
The second application implements the stores API. However,  I'm not sure whats the best way to connect this (the second) application to the first. As in,define  in the second application that creating a token depends on the token created by the first application.

I hope this  clarifies my request. Thank you.
